fuel-txpool/src/containers/dependency.rs

```diff
@@ -435,75 +439,72 @@ impl Dependency {
-                | Output::Withdrawal { .. }
-                | Output::Change { .. }
-                | Output::Variable { .. } => None,
-                Output::Contract { input_index, .. } => {
-                    if let Input::Contract { contract_id, .. } = tx.inputs()[*input_index as usize]
-                    {
-                        Some(contract_id)
+            match output {
+                Output::Withdrawal { .. } | Output::Contract { .. } => {
+                    // no other transactions can depend on these types of outputs
+                }
+                Output::Coin { .. } | Output::Change { .. } | Output::Variable { .. } => {
+                    // remove transactions that depend on this coin output
+                    let utxo = UtxoId::new(tx.id(), index as u8);
+                    if let Some(state) = self.coins.remove(&utxo).map(|c| c.is_spend_by) {
+                        // there may or may not be any dependents for this coin output
+                        if let Some(spend_by) = state {
+                            let rem_tx = txs.get(&spend_by).expect("Tx should be present in txs");
+                            removed_transactions.extend(
+                                self.recursively_remove_all_dependencies(txs, rem_tx.tx().clone()),
+                            );
+                        }
                     } else {
-                        panic!("InputIndex for contract should be always correct");
+                        // this shouldn't happen since the output belongs to this unique transaction,
+                        // no other resources should remove this coin state except this transaction.
+                        warn!("expected a coin state to be associated with {:?}", &utxo);
                     }
                 }
-                Output::ContractCreated { contract_id, .. } => Some(*contract_id),
-            };
-            // remove contract childrens.
-            if let Some(ref contract_id) = is_contract {
-                let state = self.contracts.get(contract_id).cloned();
-                if let Some(state) = state {
-                    // call removal on every tx;
-                    for rem_tx in state.used_by.iter() {
-                        let rem_tx = txs.get(rem_tx).expect("Tx should be present in txs");
-                        removed_tx.extend(
-                            self.recursively_remove_all_dependencies(txs, rem_tx.tx().clone()),
-                        );
+                Output::ContractCreated { contract_id, .. } => {
+                    // remove any other pending txs that depend on our yet to be created contract
+                    if let Some(contract) = self.contracts.remove(contract_id) {
+                        for spend_by in contract
+                            .used_by
+                            .into_iter()
+                            // skip the current tx
+                            .filter(|used_by| used_by != &id)
```

Comment:
this `.filter` is redundent as it is should not be possible that tx uses its own Output::ContractCreated

Reply:
Is there maybe a reason why you put it there?